lib/pharos/phase.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+require 'logger'
+
+module Pharos
+  class Phase
+    PHASE_TITLE = nil
+
+    # @return [String]
+    def self.title(title = nil)
+      @title = title if title
+      @title || name
+    end
+
+    def to_s
+      "#{self.class.title} @ #{@host}"
+    end
+
+    def self.register_component(component)
+      Pharos::Phases.register_component component
+    end
+
+    # @param host [Pharos::Configuration::Host]
+    # @param config [Pharos::Config]
+    # @param ssh [Pharos::SSH::Client]
+    # @param master [Pharos::Configuration::Host]
+    def initialize(host, config: nil, ssh: nil, master: nil)
+      @host = host
+      @config = config
+      @ssh = ssh
+      @master = master
```

Comment:
@kke the only thing that this `@master` is now used for is now `Pharos::Kube.client(@master.address)`:

```
lib/pharos/phases/configure_dns.rb:          @master.address,
lib/pharos/phases/configure_metrics.rb:        cert_manager = Pharos::Kube::CertManager.new(@master, 'metrics-server-client-cert', namespace: 'kube-system')
lib/pharos/phases/configure_metrics.rb:          @master.address, 'metrics-server',
lib/pharos/phases/configure_metrics.rb:        cert_manager = Pharos::Kube::CertManager.new(@master, 'heapster-client-cert', namespace: 'kube-system')
lib/pharos/phases/configure_metrics.rb:          @master.address, 'heapster',
lib/pharos/phases/configure_network.rb:        kube_client = Pharos::Kube.client(@master.address)
lib/pharos/phases/configure_network.rb:          @master.address, 'weave',
lib/pharos/phases/label_node.rb:        @kube ||= Pharos::Kube.client(@master.address)
lib/pharos/phases/store_cluster_yaml.rb:          Pharos::Kube.update_resource(@master.address, configmap)
lib/pharos/phases/store_cluster_yaml.rb:          Pharos::Kube.create_resource(@master.address, configmap)
```

That's what I mean by the `Pharos::Kube` refactoring: these class method calls with the `@master.address` should be replaced by something called `@kube` with instance methods.

Reply:
The only exception is the `ConfigureMaster#upgrade?` use of `Pharos::Kube.config_exists?(@host.address)`... I suspect that check is broken for the multi-master case, and needs some different logic anyways.